File: fors.py
from copy import deepcopy
from math import floor

from adrs import ADRS
from params import Params, PRF, H, F, Tlen
from wots import base_2b


# algorithm 14
def fors_skGen(sk_seed, pk_seed, adrs: ADRS, idx):
    """
    Generates a FORS private-key value

    Params:
        sk_seed     secret seed
        pk_seed     public seed
        ADRS        address
        idx         secret key index
    Returns:
        n-byte FORS private key
    """
    sk_adrs = deepcopy(adrs)
    sk_adrs.setTypeAndClear(Params.FORS_PRF)
    sk_adrs.setKeyPairAddress(adrs.getKeyPairAddress())
    sk_adrs.setTreeIndex(idx)
    return PRF(pk_seed, sk_seed, adrs)

#Algorithmus 15 (Computes the root of a Merkle subtree of FORS public values)
def fors_node(SK_seed, i, z, PK_seed, ADRS):    #Input: Secret seed SK.seed, target node index 𝑖, target node height 𝑧, public seed PK.seed, address ADRS

    if z == 0:
        sk = fors_skGen(SK_seed, PK_seed, ADRS, i)
        ADRS.setTreeHeight(0)
        ADRS.setTreeIndex(i)
        node = F(PK_seed, ADRS, sk)
    else:
        ADRS.setTreeHeight(z - 1)
        lnode = fors_node(SK_seed, 2 * i, z - 1, PK_seed, ADRS)
        ADRS.setTreeHeight(z - 1)
        rnode = fors_node(SK_seed, 2 * i + 1, z - 1, PK_seed, ADRS)
        ADRS.setTreeHeight(z)
        ADRS.setTreeIndex(i)

        concatenated_nodes = lnode + rnode
        node = H(PK_seed, ADRS, concatenated_nodes)

    return node                                 #Output: 𝑛-byte root 𝑛𝑜𝑑𝑒

#Algorithmus 16 (Generates a FORS signature)
def fors_sign(md, SK_seed, PK_seed, ADRS):      #Input: Message digest 𝑚𝑑, secret seed SK.seed, address ADRS, public seed PK.seed

    SIG_FORS = []                              # Initialize SIG_FORS as an empty byte string
    indices = base_2b(md, Params.a, Params.k)                 # Compute indices using base_2b function

    for i in range(Params.k):
        SIG_FORS.append(fors_skGen(SK_seed, PK_seed, ADRS, i * 2**Params.a + indices[i]))   # Compute signature elements

        AUTH = []
        for j in range(Params.a):
            s = floor(indices[i] / 2 ** j) ^ 1              # Compute auth path
            AUTH.append(fors_node(SK_seed, i * 2**(Params.a - j) + s, j, PK_seed, ADRS))
        SIG_FORS += AUTH

    # a FORS signature is:
    # (n bytes + a * n bytes ) * k (=182 for our paremeter set)
    return SIG_FORS                             # Output: FORS signature SIG𝐹 𝑂𝑅𝑆

#Algorithmus 17 (Computes a FORS public key from a FORS signature)
def fors_pkFromSig(SIG_FORS, md, PK_seed, ADRS):    # Input: FORS signature SIG𝐹 𝑂𝑅𝑆, message digest 𝑚𝑑, public seed PK.seed, address ADRS

    indices = base_2b(md, Params.a, Params.k)
    root = [0] * Params.k
    node = [0, 0]

    for i in range(Params.k):
        sk = SIG_FORS[i * (Params.a + 1):i * (Params.a + 1) + 1]    # Compute leaf
        ADRS.setTreeHeight(0)
        ADRS.setTreeIndex(i * 2**Params.a + indices[i])
        node[0] = F(PK_seed, ADRS, sk)

        auth = SIG_FORS[i * (Params.a + 1) + 1:(i + 1) * (Params.a + 1)]    # Compute root from leaf and AUTH
        for j in range(Params.a):
            ADRS.setTreeHeight(j + 1)
            if indices[i] // 2**j % 2 == 0:
                ADRS.setTreeIndex(ADRS.getTreeIndex() // 2)

                concatenated_input = node[0] + auth[j]  # Concatenate node[0] and auth[j]
                node[1] = H(PK_seed, ADRS, concatenated_input)  # Compute H(PK_seed, ADRS, node[0] || auth[j]) and store it in node[1]
            else:
                ADRS.setTreeIndex((ADRS.getTreeIndex() - 1) // 2)

                concatenated_input = auth[j] + node[0]  # Concatenate node[0] and auth[j]
                node[1] = H(PK_seed, ADRS, concatenated_input)  # Compute H(PK_seed, ADRS, node[0] || auth[j]) and store it in node[1]

            node[0] = node[1]

        root[i] = node[0]

    forspkADRS = deepcopy(ADRS)                    # Copy address to create a FORS public-key address
    forspkADRS.setTypeAndClear(Params.FORS_ROOTS)
    forspkADRS.setKeyPairAddress(ADRS.getKeyPairAddress())

    pk = Tlen(PK_seed, forspkADRS, root)          # Compute the FORS public key

    return pk                                   # Output: FORS public key

File: PostquantenCrypto/fors.c
#include <string.h>
#include <math.h>
#include <stdlib.h>
#include "params.h"
#include "adrs.h"
#include "shake.h"
#include "wots.h"

// algorithm 14
unsigned char* fors_skGen(const unsigned char* sk_seed, const unsigned char* pk_seed,
                         ADRS* adrs, int idx) {
    /*
    Generates a FORS private-key value

    Params:
        sk_seed     secret seed
        pk_seed     public seed
        ADRS        address
        idx         secret key index
    Returns:
        n-byte FORS private key
    */
    ADRS sk_adrs;
    memcpy(&sk_adrs, adrs, sizeof(ADRS));
    setTypeAndClear(&sk_adrs, FORS_PRF);
    setKeyPairAddress(&sk_adrs, getKeyPairAddress(adrs));
    setTreeIndex(&sk_adrs, idx);
    return PRF(pk_seed, sk_seed, &sk_adrs);
}

// Algorithm 15 (Computes the root of a Merkle subtree of FORS public values)
unsigned char* fors_node(const unsigned char* SK_seed, int i, int z,
                        const unsigned char* PK_seed, ADRS* ADRS) {
    unsigned char* node;
    node = malloc(N_BYTES);  // Assuming N_BYTES is the size of hash output

    if (z == 0) {
        unsigned char* sk = fors_skGen(SK_seed, PK_seed, ADRS, i);
        setTreeHeight(ADRS, 0);
        setTreeIndex(ADRS, i);
        F(PK_seed, ADRS, sk, node);
        free(sk);
    } else {
        unsigned char* lnode;
        unsigned char* rnode;
        unsigned char* concatenated_nodes;

        setTreeHeight(ADRS, z - 1);
        lnode = fors_node(SK_seed, 2 * i, z - 1, PK_seed, ADRS);

        setTreeHeight(ADRS, z - 1);
        rnode = fors_node(SK_seed, 2 * i + 1, z - 1, PK_seed, ADRS);

        setTreeHeight(ADRS, z);
        setTreeIndex(ADRS, i);

        concatenated_nodes = malloc(2 * N_BYTES);
        memcpy(concatenated_nodes, lnode, N_BYTES);
        memcpy(concatenated_nodes + N_BYTES, rnode, N_BYTES);

        H(PK_seed, ADRS, concatenated_nodes, node);

        free(lnode);
        free(rnode);
        free(concatenated_nodes);
    }

    return node;
}

// Algorithm 16 (Generates a FORS signature)
unsigned char** fors_sign(const unsigned char* md, const unsigned char* SK_seed,
                         const unsigned char* PK_seed, ADRS* ADRS) {
    unsigned char** SIG_FORS;
    int* indices;
    int total_elements = K * (A + 1);  // K and A from params

    SIG_FORS = malloc(total_elements * sizeof(unsigned char*));
    indices = base_2b(md, A, K);

    for (int i = 0; i < K; i++) {
        SIG_FORS[i * (A + 1)] = fors_skGen(SK_seed, PK_seed, ADRS,
                                          i * (1 << A) + indices[i]);

        for (int j = 0; j < A; j++) {
            int s = (floor(indices[i] / pow(2, j))) ^ 1;
            SIG_FORS[i * (A + 1) + j + 1] = fors_node(SK_seed,
                                                     i * (1 << (A - j)) + s,
                                                     j, PK_seed, ADRS);
        }
    }

    free(indices);
    return SIG_FORS;
}

// Algorithm 17 (Computes a FORS public key from a FORS signature)
unsigned char* fors_pkFromSig(unsigned char** SIG_FORS, const unsigned char* md,
                            const unsigned char* PK_seed, ADRS* ADRS) {
    int* indices = base_2b(md, A, K);
    unsigned char** root = malloc(K * sizeof(unsigned char*));
    unsigned char* node[2];
    unsigned char* pk;

    node[0] = malloc(N_BYTES);
    node[1] = malloc(N_BYTES);

    for (int i = 0; i < K; i++) {
        unsigned char* sk = SIG_FORS[i * (A + 1)];
        setTreeHeight(ADRS, 0);
        setTreeIndex(ADRS, i * (1 << A) + indices[i]);
        F(PK_seed, ADRS, sk, node[0]);

        for (int j = 0; j < A; j++) {
            unsigned char* auth = SIG_FORS[i * (A + 1) + j + 1];
            unsigned char* concatenated_input = malloc(2 * N_BYTES);

            setTreeHeight(ADRS, j + 1);

            if ((indices[i] / (1 << j)) % 2 == 0) {
                setTreeIndex(ADRS, getTreeIndex(ADRS) / 2);
                memcpy(concatenated_input, node[0], N_BYTES);
                memcpy(concatenated_input + N_BYTES, auth, N_BYTES);
            } else {
                setTreeIndex(ADRS, (getTreeIndex(ADRS) - 1) / 2);
                memcpy(concatenated_input, auth, N_BYTES);
                memcpy(concatenated_input + N_BYTES, node[0], N_BYTES);
            }

            H(PK_seed, ADRS, concatenated_input, node[1]);
            memcpy(node[0], node[1], N_BYTES);

            free(concatenated_input);
        }

        root[i] = malloc(N_BYTES);
        memcpy(root[i], node[0], N_BYTES);
    }

    ADRS forspkADRS;
    memcpy(&forspkADRS, ADRS, sizeof(ADRS));
    setTypeAndClear(&forspkADRS, FORS_ROOTS);
    setKeyPairAddress(&forspkADRS, getKeyPairAddress(ADRS));

    pk = Tlen(PK_seed, &forspkADRS, root);

    // Cleanup
    free(node[0]);
    free(node[1]);
    for (int i = 0; i < K; i++) {
        free(root[i]);
    }
    free(root);
    free(indices);

    return pk;
}
